src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,212 @@
+  private prices: {
+    [platform: string]: {
+      [currency: string]: PriceCache;
+    };
+  } = {};
+
+  protected constructor(protected logger: Logger) {}
+
+  public static get(logger: Logger): PriceClient {
+    if (!this.instance) this.instance = new PriceClient(logger);
+
+    return this.instance;
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  // @todo: Could be nice to specify a priority rather than relying on order of addition.
+  addPriceFeed(priceFeed: PriceFeedAdapter): void {
+    assert(!this.priceFeeds.includes(priceFeed), `Price feed ${priceFeed.name} already registered.`);
+    this.priceFeeds.push(priceFeed);
+    this.logger.debug({
+      at: "PriceClient#addPriceFeed",
+      message: `Appended new price feed: ${priceFeed.name}.`,
+    });
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  clearPriceFeeds(): void {
+    const cleared: string[] = this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+    this.priceFeeds = [];
+    this.logger.debug({
+      at: "PriceClient#clearPriceFeeds",
+      message: "Cleared all price feeds.",
+      cleared,
+    });
+  }
+
+  async getTokenPrice(address: string, currency = "usd", platform = "ethereum"): Promise<TokenPrice> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache: PriceCache = this.getPriceCache(currency, platform);
+    const now: number = msToS(Date.now());
+
+    let tokenPrice: TokenPrice | undefined = priceCache[address];
+    const cacheMiss = tokenPrice === undefined || now - this.maxPriceAge > tokenPrice.timestamp;
+
+    if (this.maxPriceAge > 0) {
+      const age: number = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_VALUE;
+      this.logger.debug({
+        at: "PriceClient#getTokenPrice",
+        message: `Cache ${cacheMiss ? "miss" : "hit"} on ${platform}/${currency} for token ${address}.`,
+        age: `${age} S`,
+        price: tokenPrice,
+      });
+    }
+
+    if (cacheMiss) {
+      const prices: TokenPrice[] = await this.getTokenPrices([address], currency, platform);
+      tokenPrice = prices[0];
+    }
+    return tokenPrice;
+  }
+
+  async getTokenPrices(addresses: string[], currency = "usd", platform = "ethereum"): Promise<TokenPrice[]> {
+    assert(this.priceFeeds.length > 0, "No price feeds were registerted.");
+    const priceCache: PriceCache = this.getPriceCache(currency, platform);
+
+    // Pre-populate price cache with requested token addresses
+    this.initPrices(priceCache, addresses);
+
+    const prices: PriceCache = await this.requestPrices(addresses, currency, platform);
```

Comment:
Might make sense to guard this by first checking all timestamps, and only proceeding to query if at least one of them are out of date.